lib/TPP/TileConsumerAndFuseProducers.cpp

```diff
@@ -25,108 +27,92 @@ using namespace mlir;
-                                               OpBuilder &builder) {
-  if (linalgOp.getNumLoops() != tiles.size())
-    return failure();
-  SmallVector<Range> iterationDomain =
-      cast<TilingInterface>(linalgOp.getOperation())
-          .getIterationDomain(builder);
-  for (const auto &it : llvm::enumerate(iterationDomain)) {
-    // fine, we are not tiling along this dimension.
-    if (tiles[it.index()] == 0)
-      continue;
-    // require static loop range
-    if (!isStaticRange(it.value()))
-      return failure();
-    int64_t sizeRange = getSizeRange(it.value());
-    // fail if the tail size equals the range
-    // or is not a full tile.
-    if (tiles[it.index()] == sizeRange)
-      return failure();
-    if (sizeRange % tiles[it.index()] != 0)
-      return failure();
+static bool hasMatmulLikeProducer(linalg::LinalgOp linalgOp) {
+  for (Value operand : linalgOp->getOperands()) {
+    Operation *op = operand.getDefiningOp<linalg::LinalgOp>();
+    if (op && isMatmulLike(op))
+      return true;
   }
-  return success();
+  return false;
 }
 
-struct FuseGenericOp : public OpRewritePattern<linalg::GenericOp> {
-  FuseGenericOp(MLIRContext *context, ArrayRef<int64_t> tileSizes,
-                PatternBenefit benefit = 1)
-      : OpRewritePattern<linalg::GenericOp>(context, benefit),
-        tileSizes(tileSizes) {}
+static FailureOr<scf::SCFTilingResult>
+tileConsumer(RewriterBase &rewriter, TilingInterface consumer,
+             ArrayRef<int64_t> tileSizes) {
+  // If tileSizes is provided use them otherwise defaul to 1, 1, meaning fuse
+  // entire the two outermost loops.
+  SmallVector<int64_t> actualTileSizes = tileSizes.empty()
+                                             ? SmallVector<int64_t>{1, 1}
+                                             : llvm::to_vector(tileSizes);
+  auto options = scf::SCFTilingOptions().setTileSizes(actualTileSizes);
+  FailureOr<scf::SCFTilingResult> tilingResult =
+      scf::tileUsingSCFForOp(rewriter, consumer, options);
+  return tilingResult;
+}
 
-  bool isInplace(linalg::LinalgOp linalgOp) const {
-    return linalgOp.getNumDpsInputs() == 0;
+// Tile and fuse a matmul along the parallel dimensions.
+static FailureOr<scf::SCFTileAndFuseResult>
+fuseMatmulLikeAndEltwise(RewriterBase &rewriter, TilingInterface consumer,
+                         ArrayRef<int64_t> tileSizes) {
+
+  // Step 1. tile the consumer.
+  scf::SCFTileAndFuseResult tileAndFuseResult;
+  FailureOr<scf::SCFTilingResult> tilingResult =
+      tileConsumer(rewriter, consumer, tileSizes);
+  if (failed(tilingResult))
+    return rewriter.notifyMatchFailure(consumer,
+                                       "failed to tile base operation");
+  for (auto *tiledOp : tilingResult->tiledOps)
+    tileAndFuseResult.tiledAndFusedOps.insert(tiledOp);
+  tileAndFuseResult.loops = std::move(tilingResult->loops);
+  for (const auto &result : llvm::enumerate(
```

Comment:
Use auto tuple notation to avoid `get<N>` and `zip_equal`.

Reply:
C++ 17 does not support nesting in structured bindings. But I can replace llvm::zip with llvm::zip_equal